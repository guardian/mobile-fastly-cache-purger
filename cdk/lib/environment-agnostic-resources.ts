import {Repository, TagMutability} from "aws-cdk-lib/aws-ecr";
import * as iam from "aws-cdk-lib/aws-iam";
import {GuardianAwsAccounts} from "@guardian/private-infrastructure-config";
import {GuStack, GuStackProps} from "@guardian/cdk/lib/constructs/core";
import {App} from "aws-cdk-lib";

export class EnvironmentAgnosticResources extends GuStack {
    constructor(scope: App, id: string, props: GuStackProps) {
        super(scope, id, props);
        const ecrRepository = new Repository(this, 'mobile-fastly-cache-purger-repo', {
            repositoryName: 'mobile-fastly-cache-purger',
            imageScanOnPush: true,
            imageTagMutability: TagMutability.IMMUTABLE
        })
        new iam.Role(this, 'CIRole', {
            assumedBy:  new iam.FederatedPrincipal(
                `arn:aws:iam::${GuardianAwsAccounts.Mobile}:oidc-provider/token.actions.githubusercontent.com`,
                {
                    "StringEquals": { "token.actions.githubusercontent.com:aud": "sts.amazonaws.com" },
                    "StringLike": { "token.actions.githubusercontent.com:sub": "repo:guardian/mobile-fastly-cache-purger:*"}
                },
                "sts:AssumeRoleWithWebIdentity"
            ),
            inlinePolicies: {
                ecrToken: new iam.PolicyDocument({
                    statements: [
                        new iam.PolicyStatement({
                            actions: ['ecr:GetAuthorizationToken'],
                            resources: ['*']
                        })
                    ]
                }),
                ecrUpload: new iam.PolicyDocument({
                    statements: [
                        new iam.PolicyStatement({
                            actions: [
                                'ecr:CompleteLayerUpload',
                                'ecr:UploadLayerPart',
                                'ecr:InitiateLayerUpload',
                                'ecr:BatchCheckLayerAvailability',
                                'ecr:PutImage'
                            ],
                            resources: [ecrRepository.repositoryArn]
                        })
                    ]
                })

            }
        })
    }
}


