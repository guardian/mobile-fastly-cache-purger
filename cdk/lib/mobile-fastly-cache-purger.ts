import type {GuStackProps} from '@guardian/cdk/lib/constructs/core';
import {GuStack} from '@guardian/cdk/lib/constructs/core';
import {GuLambdaFunction} from '@guardian/cdk/lib/constructs/lambda';
import {GuardianAwsAccounts} from '@guardian/private-infrastructure-config';
import type {App} from 'aws-cdk-lib';
import {CfnParameter, Duration, Fn} from 'aws-cdk-lib';
// eslint-disable-next-line import/no-namespace -- sdf
import * as iam from 'aws-cdk-lib/aws-iam';
import {Runtime} from 'aws-cdk-lib/aws-lambda';
// eslint-disable-next-line import/no-namespace -- sdf
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import {Topic} from 'aws-cdk-lib/aws-sns';
import {SqsSubscription} from "aws-cdk-lib/aws-sns-subscriptions";
// eslint-disable-next-line import/no-namespace -- sdf
import * as sqs from 'aws-cdk-lib/aws-sqs';
import {GuLambdaDockerFunction} from "./gu-lamba-docker-function";


export class MobileFastlyCachePurger extends GuStack {

	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const faciaID = this.stage == "CODE" ? "StorageConsumerRole-1JWVQ2NTELFT7" : "StorageConsumerRole-1R9GQEVJIM323";
		const buildId = new CfnParameter(this, 'BuildId', {
			type: 'String',
			default: 'dev',
			description: 'Tag to be used for the image URL, e.g. riff raff build id',
		}).value.toString();

		const executionRole: iam.Role = new iam.Role(this, 'ExecutionRole', {
			assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
			path: "/",
			inlinePolicies: {
				logs: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: [ 'logs:CreateLogGroup' ],
							resources: [ `arn:aws:logs:eu-west-1:${this.account}:*` ]
						}),
						new iam.PolicyStatement({
							actions: [ 'logs:CreateLogStream', 'logs:PutLogEvents' ],
							resources: [ `arn:aws:logs:eu-west-1:${this.account}:log-group:/aws/lambda/*:*` ]
						})
					] }),
				Conf: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: [ 'ssm:GetParametersByPath' ],
							resources: [ `arn:aws:ssm:${this.region}:${this.account}:parameter/cache-purger/${this.stage}` ]
						})
					] }),
				Assume: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: ['sts:AssumeRole'],
							resources: [`arn:aws:iam::${GuardianAwsAccounts.CMSFronts}:role/facia-${this.stage}-${faciaID}`]
						})
					]
				})
			}
		})

		const handler: GuLambdaFunction = new GuLambdaFunction(this, 'mobile-fastly-cache-purger', {
			handler: 'PurgerLambda::handleRequest',
			functionName: `mobile-fastly-cache-purger-cdk-${this.stage}`,
			timeout: Duration.seconds(60),
			environment: {
				App: 'mobile-fastly-cache-purger',
				Stack: this.stack,
				Stage: this.stage,
			},
			runtime: Runtime.JAVA_11,
			app: 'mobile-fastly-cache-purger',
			fileName: `mobile-fastly-cache-purger.jar`,
			role: executionRole,
		});

		const imageRepositoryArn = Fn.importValue('mobile-fastly-cache-purger-repository-arn')
		const imageRepositoryName = Fn.importValue('mobile-fastly-cache-purger-repository-name')
		const lambdaFunction: GuLambdaDockerFunction = new GuLambdaDockerFunction(this, 'mobile-fastly-cache-purger-v2', {
			functionName: `mobile-fastly-cache-purger-cdk-${this.stage}-v2`,
			timeout: Duration.seconds(60),
			environment: {
				App: 'mobile-fastly-cache-purger-v2',
				Stack: this.stack,
				Stage: this.stage,
			},
			app: 'mobile-fastly-cache-purger-v2',
			repositoryArn: `${imageRepositoryArn}`,
			repositoryName: `${imageRepositoryName}`,
			imageTag: `${buildId}`,
			role: executionRole,
			memorySize: 1024
		});

		const dlq: sqs.Queue = new sqs.Queue(this, "frontsPurgeDlq")

		const queue: sqs.Queue = new sqs.Queue(this, "frontsPurgeSqs", {
			visibilityTimeout: Duration.seconds(70), 	//default for a queue is 30s, and the lambda is 60s
			deadLetterQueue: {
				maxReceiveCount: 3,
				queue: dlq,
			}
		});

		const frontsUpdateTopicName=
			this.stage == "CODE" ? "FrontsUpdateSNSTopic-RepwK3g95V3w" : "FrontsUpdateSNSTopic-kWN6oX2kvOmI";

		const frontsUpdateTopic = Topic.fromTopicArn(
			this,
			"FrontsUpdateSNSTopic",
			`arn:aws:sns:${this.region}:${GuardianAwsAccounts.CMSFronts}:facia-${this.stage}-${frontsUpdateTopicName}`
		)
		frontsUpdateTopic.addSubscription(new SqsSubscription(queue));

		const eventSource: lambdaEventSources.SqsEventSource = new lambdaEventSources.SqsEventSource(queue);

		handler.addEventSource(eventSource);
		lambdaFunction.addEventSource(eventSource);
	}
}
