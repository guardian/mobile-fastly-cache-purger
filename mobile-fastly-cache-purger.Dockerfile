FROM public.ecr.aws/lambda/java:11
COPY mobile-fastly-cache-purger.jar ${LAMBDA_TASK_ROOT}/lib/
CMD ["PurgerLambda::handleRequest"]
