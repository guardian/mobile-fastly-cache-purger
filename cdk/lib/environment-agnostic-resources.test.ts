import {App} from "aws-cdk-lib";
import {MobileFastlyCachePurger} from "./mobile-fastly-cache-purger";
import {Template} from "aws-cdk-lib/assertions";
import {EnvironmentAgnosticResources} from "./environment-agnostic-resources";

describe('The EnvironmentAgnosticResources stack', () => {
    it('matches the snapshot', () => {
        const app = new App();
        const stack = new EnvironmentAgnosticResources(app, 'EnvironmentAgnosticResources', {
            stack: 'mobile-fastly-cache-purger',
            stage: 'TEST',
        });
        const template = Template.fromStack(stack);
        expect(template.toJSON()).toMatchSnapshot();
    });
})
