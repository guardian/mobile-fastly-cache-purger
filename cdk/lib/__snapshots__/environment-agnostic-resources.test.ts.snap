// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EnvironmentAgnosticResources stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mobilefastlycachepurgerrepositoryarn": {
      "Export": {
        "Name": "mobile-fastly-cache-purger-repository-arn",
      },
      "Value": {
        "Fn::GetAtt": [
          "mobilefastlycachepurgerrepo79A8A316",
          "Arn",
        ],
      },
    },
    "mobilefastlycachepurgerrepositoryname": {
      "Export": {
        "Name": "mobile-fastly-cache-purger-repository-name",
      },
      "Value": {
        "Fn::GetAtt": [
          "mobilefastlycachepurgerrepo79A8A316",
          "Arn",
        ],
      },
    },
  },
  "Resources": {
    "CIRole1307643F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:guardian/mobile-fastly-cache-purger:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "arn:aws:iam::000000000033:oidc-provider/token.actions.githubusercontent.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "ecr:GetAuthorizationToken",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ecrToken",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ecr:CompleteLayerUpload",
                    "ecr:UploadLayerPart",
                    "ecr:InitiateLayerUpload",
                    "ecr:BatchCheckLayerAvailability",
                    "ecr:PutImage",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "mobilefastlycachepurgerrepo79A8A316",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ecrUpload",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilefastlycachepurgerrepo79A8A316": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "ImageScanningConfiguration": {
          "ScanOnPush": true,
        },
        "ImageTagMutability": "IMMUTABLE",
        "RepositoryName": "mobile-fastly-cache-purger",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
